src/many-identity-dsa/src/lib.rs

```diff
@@ -0,0 +1,171 @@
+
+            #[cfg(feature = "ecdsa")]
+            CoseKeyImpl::EcDsa(i) => i.address(),
+
+            CoseKeyImpl::Illegal_ => unreachable!(),
+        }
+    }
+
+    pub fn public_key(&self) -> Option<CoseKey> {
+        match self {
+            #[cfg(feature = "ed25519")]
+            CoseKeyImpl::Ed25519(i) => Identity::public_key(i),
+
+            #[cfg(feature = "ecdsa")]
+            CoseKeyImpl::EcDsa(i) => Identity::public_key(i),
+
+            CoseKeyImpl::Illegal_ => unreachable!(),
+        }
+    }
+
+    pub fn sign_1(&self, envelope: CoseSign1) -> Result<CoseSign1, ManyError> {
+        match self {
+            #[cfg(feature = "ed25519")]
+            CoseKeyImpl::Ed25519(i) => i.sign_1(envelope),
+
+            #[cfg(feature = "ecdsa")]
+            CoseKeyImpl::EcDsa(i) => i.sign_1(envelope),
+
+            CoseKeyImpl::Illegal_ => unreachable!(),
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct CoseKeyIdentity {
+    inner: CoseKeyImpl,
+}
+
+impl Debug for CoseKeyIdentity {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_tuple("CoseKeyIdentity")
+            .field(&self.address())
+            .finish()
+    }
+}
+
+impl CoseKeyIdentity {
+    pub fn from_pem(pem: impl AsRef<str>) -> Result<Self, ManyError> {
+        Ok(Self {
+            inner: CoseKeyImpl::from_pem(pem.as_ref())
+                .ok_or_else(|| ManyError::unknown("Algorithm unsupported."))?,
+        })
+    }
+}
+
+impl Identity for CoseKeyIdentity {
+    fn address(&self) -> Address {
+        self.inner.address()
+    }
+
+    fn public_key(&self) -> Option<CoseKey> {
+        self.inner.public_key()
+    }
+
+    fn sign_1(&self, envelope: CoseSign1) -> Result<CoseSign1, ManyError> {
+        self.inner.sign_1(envelope)
+    }
+}
+
+macro_rules! try_verify {
+    ($init: expr, $envelope: ident, $name: literal) => {
+        match $init {
+            Ok(v) => {
+                return v.sign_1($envelope);
+            }
+            Err(err) => {
+                trace!("Initialization error ({}): {}", $name, err)
+            }
+        }
+    };
+}
+
+#[derive(Clone)]
+pub struct CoseKeyVerifier;
+
+impl Verifier for CoseKeyVerifier {
+    fn sign_1(&self, envelope: &CoseSign1) -> Result<(), ManyError> {
+        let keyid = &envelope.protected.header.key_id;
+
+        // Extract the keyset argument.
+        let keyset = keyset_from_cose_sign1(envelope)
+            .ok_or_else(|| ManyError::unknown("Could not find keyset in headers."))?;
+
+        let key = keyset
+            .0
+            .iter()
+            .find(|key| key.key_id.eq(keyid))
+            .ok_or_else(|| ManyError::unknown("Could not find the key in keyset."))?;
+
+        #[cfg(feature = "ed25519")]
+        try_verify!(ed25519::Ed25519Verifier::from_key(key), envelope, "ed25519");
+
+        #[cfg(feature = "ecdsa")]
+        try_verify!(ecdsa::EcDsaVerifier::from_key(key), envelope, "ecdsa");
```

Comment:
Yeah. This will complicate the code later if we add more EdDSA curves, for example. Unless we can see this is a bottleneck I'd like to defer whether or not a key is supported to its implementation, for now.

Reply:
For the sake of posterity, the `CoseKey` structure should contain all this information already (key type, algorithm, curve type, etc.).

See https://datatracker.ietf.org/doc/html/rfc8152#page-41 for example.